test/irmin-pack/test_existing_stores.ml

```
@@ -286,6 +286,78 @@ module Test_corrupted_stores = struct
+      S.Commit.v repo ~parents:[] ~info:(info ()) tree
+    in
+    let check_commit repo commit k v =
+      commit |> S.Commit.hash |> S.Commit.of_hash repo >>= function
+      | None ->
+          Alcotest.failf "Commit `%a' is dangling in repo" S.Commit.pp_hash
+            commit
+      | Some commit ->
+          let tree = S.Commit.tree commit in
+          S.Tree.find tree k
+          >|= Alcotest.(check (option string))
+                (Fmt.strf "Expected binding [%a ↦ %s]"
+                   Fmt.(Dump.list string)
+                   k v)
+                (Some v)
+    in
+    let check_upper repo msg exp =
+      let got = S.PrivateLayer.upper_in_use repo in
+      let cast x = (x :> [ `Upper0 | `Upper1 | `Lower ]) in
+      if not (got = exp) then
+        Alcotest.failf "%s expected %a got %a" msg Irmin_layers.pp_layer_id
+          (cast exp) Irmin_layers.pp_layer_id (cast got)
+    in
+    let* rw = S.Repo.v (config ~fresh:false root) in
+    let* ro = S.Repo.v (config ~fresh:false ~readonly:true root) in
+    Log.app (fun l -> l "Open a layered store aborted during a freeze");
+    Alcotest.(check bool) "Store needs recovery" (S.needs_recovery rw) true;
+    check_upper rw "Upper before recovery" `Upper1;
+    add_commit rw [ "a" ] "x" >>= fun c ->
+    S.sync ro;
+    check_commit ro c [ "a" ] "x" >>= fun () ->
+    Log.app (fun l -> l "Freeze with recovery flag set");
+    S.freeze ~recovery:true ~max:[ c ] rw >>= fun () ->
+    S.PrivateLayer.wait_for_freeze () >>= fun () ->
+    Alcotest.(check bool)
+      "Store doesn't need recovery" (S.needs_recovery rw) false;
+    check_upper rw "Upper after recovery" `Upper0;
+    check_commit rw c [ "a" ] "x" >>= fun () ->
+    S.sync ro;
+    check_commit ro c [ "a" ] "x" >>= fun () ->
+    check_upper ro "RO upper after recovery" `Upper0;
+    add_commit rw [ "b" ] "y" >>= fun c ->
+    Log.app (fun l ->
+        l "Freeze ignores the recovery flag when it doesn't need recovery ");
```

I have a feeling we've discussed this before, but why do we want the user to determine when to do a recovery if the store knows this? It seems roundabout to have a `needs_recovery` function if we could just automatically recover.

With the current API, I think passing `~recovery:true` should either raise an exception or do the recovery anyway (not just be silently ignored).

---

The store can tell that a freeze was interrupted, but it cannot infer the rest of arguments, i.e. the max commits. So I think we should have a freeze with an explicit recovery flag, which might need different arguments than the normal freeze. What do you think?
I remove the silent ignore, the freeze now always does the recovery when the flag is set.

---

That makes sense, thanks.